packages/utils/src/filterEntities.js

```diff
@@ -0,0 +1,71 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+const FILTER_TYPE_TO_METHOD = {
+  '=': (entity, field, filterValue) => entity[field] === filterValue,
+  in: (entity, field, filterValue) => filterValue.includes(entity[field]),
+};
+
+const applyFilter = (entities, field, operator, value) => {
+  const filterMethod = FILTER_TYPE_TO_METHOD[operator];
+  return filterMethod ? entities.filter(entity => filterMethod(entity, field, value)) : entities;
+};
+
+const filterByObjectField = (entities, field, filterValue) => {
+  if (typeof filterValue !== 'object') {
+    throw new Error(`Filtering by ${field} expects an object`);
+  }
+
+  let filteredEntities = entities;
+  Object.entries(filterValue).forEach(([key, value]) => {
+    filteredEntities = filteredEntities.filter(e => e[field][key] === value);
+  });
+  return filteredEntities;
+};
+
+const filterByPlainField = (entities, field, filterValue, operator = '=') => {
+  if (typeof filterValue === 'object') {
+    let filteredEntities = entities;
+
+    Object.entries(filterValue).forEach(([individualFilterOperator, individualFilterValue]) => {
+      filteredEntities = applyFilter(
+        filteredEntities,
+        field,
+        individualFilterOperator,
+        individualFilterValue,
+      );
+    });
+
+    return filteredEntities;
+  }
+
+  return applyFilter(entities, field, operator, filterValue);
+};
+
+/**
+ * Example entityFilter format:
+ * ```
+ * {
+ *    code: 'ABC'
+ *    attributes: { type: 'Primary' },
+ * }
+ * ```
+ */
+export const filterEntities = (entities, entityFilter) => {
+  let filteredEntities = entities;
```

Comment:
I generalised this function a bit since it is now part of `@tupaia/utils`. We could probably make this even more generic, something like:

`export const filterObjects = (objects, filter)`

The `switch` block below explicitly uses different filter logic (plain/object) for different field keys. In practice, we don't need to be aware of the object's schema: we can figure out if a field is an object by checking its value.

Also, I have the suspicion that `FILTER_TYPE_TO_METHOD` has some commonalities with `checkValueSatisfiesCondition`.

Finally, we probably can avoid using `let`, and rather use chained array methods (`every`/`filter` etc).

I decided to not do all this refactoring since it felt out of scope for this card, but now that I think of it we're gonna do a regression testing of this functionality anyway, so it may be worth it! But if I do, perhaps it should be done as a separate PR for ease of reviewing. I would love a second opinion so please let me know!